Source/Core/DolphinQt2/Config/GeckoWidget.cpp

```diff
@@ -0,0 +1,169 @@
+
+  m_gecko_codes = Gecko::LoadCodes(m_game_ini_default, m_game_ini_local);
+
+  UpdateList();
+}
+
+void GeckoWidget::CreateWidgets()
+{
+  auto* layout = new QGridLayout;
+  m_warning = new CheatWarningWidget(m_game_id);
+  m_code_list = new QListWidget;
+  m_code_description = new QTextEdit;
+  m_code_view = new QListWidget;
+  m_download_codes = new QPushButton(tr("Download Codes (WiiRD Database)"));
+
+  m_code_description->setTextInteractionFlags(Qt::TextBrowserInteraction);
+
+  m_download_codes->setEnabled(!m_game_id.empty());
+
+  layout->addWidget(m_warning, 0, 0);
+  layout->addWidget(m_code_list, 1, 0);
+  layout->addWidget(m_code_description, 2, 0);
+  layout->addWidget(m_code_view, 3, 0);
+  layout->addWidget(m_download_codes, 4, 0);
+
+  setLayout(layout);
+}
+
+void GeckoWidget::ConnectWidgets()
+{
+  connect(m_code_list, &QListWidget::itemSelectionChanged, this, &GeckoWidget::OnSelectionChanged);
+  connect(m_code_list, &QListWidget::itemChanged, this, &GeckoWidget::OnItemChanged);
+
+  connect(m_download_codes, &QPushButton::pressed, this, &GeckoWidget::DownloadCodes);
+
+  connect(m_warning, &CheatWarningWidget::OpenGeneralSettings, this,
+          &GeckoWidget::OpenGeneralSettings);
+}
+
+void GeckoWidget::OnSelectionChanged()
+{
+  auto items = m_code_list->selectedItems();
+
+  if (items.empty())
+    return;
+
+  auto* selected = items[0];
+
+  auto& code = m_gecko_codes[selected->data(ENTRY_INDEX).toInt()];
+
+  QString notes;
+
+  for (const std::string& line : code.notes)
+    notes.append(QString::fromStdString(line));
+
+  m_code_description->setText(
+      tr("Creator: %1\nDescription:\n\n%2")
+          .arg(QString::fromStdString(code.creator), notes.isEmpty() ? tr("N/A") : notes));
+
+  m_code_view->clear();
+
+  for (const auto& c : code.codes)
+    m_code_view->addItem(QString::asprintf("%08X  %08X", c.address, c.data));
+}
+
+void GeckoWidget::OnItemChanged(QListWidgetItem* item)
+{
+  // TODO: Implement Cheat Warning Panel
+
+  m_gecko_codes[item->data(ENTRY_INDEX).toInt()].enabled = (item->checkState() == Qt::Checked);
+  Gecko::SaveCodes(m_game_ini_local, m_gecko_codes);
+
+  auto rev_filename = m_game_id + std::to_string(m_game_revision) + ".ini";
+
+  auto filename = File::Exists(File::GetUserPath(D_GAMESETTINGS_IDX) + rev_filename) ?
+                      rev_filename :
+                      m_game_id + ".ini";
+
+  m_game_ini_local.Save(File::GetUserPath(D_GAMESETTINGS_IDX) + filename);
+}
+
+void GeckoWidget::UpdateList()
+{
+  m_code_list->clear();
+
+  for (quint64 i = 0; i < m_gecko_codes.size(); i++)
+  {
+    auto& code = m_gecko_codes[i];
+
+    auto* item = new QListWidgetItem(QString::fromStdString(code.name));
+
+    item->setFlags(Qt::ItemIsEnabled | Qt::ItemIsSelectable | Qt::ItemIsUserCheckable);
+    item->setCheckState(code.enabled ? Qt::Checked : Qt::Unchecked);
+
+    // Encode data into entries for ease of access
+    item->setData(ENTRY_INDEX, i);
```

Comment:
Why store the indices in the data role, when they will always match between the two arrays? This seems overly-complicated to me.

Creating a `QAbstractItemModel` wrapping the `m_gecko_codes` array would be a better way to express the pattern "Each item displayed is associated with one item in this array" (and makes it easy to e.g. sort them). Otherwise, I'd be fine with this code just assuming that the two arrays always match perfectly for now.